File: testdata/polyglot/ruby/vuln.rb
# Deliberately vulnerable appsec fixture. Every issue is planted. DO NOT fix.
#
# A Sinatra handler so semgrep's taint rules see `params` as a tainted source.
require "sinatra"

get "/vuln" do
  # PLANT(rb-sqli, min-profile=standard, CWE-89): SQL injection via interpolated request param
  User.where("name = '#{params[:name]}'")

  # PLANT-GAP: OS command injection via system() with request param (CWE-78) — semgrep reports the eval-family CWE-94 on this file instead
  system("ping -c 1 #{params[:host]}")

  # PLANT(rb-code-injection, min-profile=standard, CWE-94): arbitrary code execution via eval of request param (semgrep emits CWE-94)
  eval(params[:code])

  # PLANT(rb-deser, min-profile=standard, CWE-502): insecure deserialization of request data
  Marshal.load(params[:data])

  # PLANT(rb-xss, min-profile=standard, CWE-79): reflected XSS via unescaped request param in HTML
  "<h1>Hello, #{params[:greeting]}</h1>"
end

File: testdata/polyglot/php/vuln.php
<?php
// Deliberately vulnerable appsec fixture. Every issue is planted. DO NOT fix.

// PLANT(php-sqli, min-profile=standard, CWE-89): SQL injection via string concatenation in mysqli_query
$conn = new mysqli("localhost", "root", "", "testdb");
$id = $_GET['id'];
$sql = "SELECT * FROM users WHERE id = " . $id;
mysqli_query($conn, $sql);

// PLANT(php-cmdi, min-profile=standard, CWE-78): OS command injection via system() with unescaped user input
$host = $_GET['host'];
system("ping -c 4 " . $host);

// PLANT(php-xss, min-profile=standard, CWE-79): reflected XSS by echoing raw GET parameter into HTML context
$name = $_GET['name'];
echo "<h1>Welcome, " . $name . "</h1>";

// PLANT(php-lfi, min-profile=max, CWE-98): local file inclusion via unvalidated user input in include() (textbook CWE-22; semgrep emits CWE-98)
$file = $_GET['file'];
include($file);
?>
